File: daily-exercise/special_topic/dp/longest-increasing-subsequence-number.go
func findNumberOfLIS(nums []int) int {
    n := len(nums)
    if n == 1{
        return n
    }
    dpLength := make([]int,n)
    dpCount := make([]int,n)
    for i:=0;i<n;i++{
        dpLength[i] = 1
        dpCount[i] = 1
    }
    maxLength := 1
    for i:=1;i<n;i++{
        for j:=0;j<i;j++{
            if nums[i] > nums[j] {
                if dpLength[j] + 1 > dpLength[i] {
                    dpLength[i] = dpLength[j]+1
                    dpCount[i] = dpCount[j]
                } else if dpLength[j] + 1 == dpLength[i] {
                    dpCount[i] += dpCount[j]
                }
            }
        }
        maxLength = max(maxLength,dpLength[i])
    }
    maxCount := 0
    for i:=0;i<n;i++{
        if dpLength[i] == maxLength{
            maxCount += dpCount[i]
        }
    }
    return maxCount
}

func max(a,b int) int {
    if a>b{
        return a
    }
    return b
}

File: daily-exercise/special_topic/dp/longest-increasing-subsequence-number.py
# -*- coding: utf-8 -*-
# @Time    : 2022/8/8 21:26
# @Author  : kevin
# @Version : python 3.7
# @Desc    : number-of-longest-increasing-subsequence
# 最大上升子序列个数
class Solution:
    def findNumberOfLIS(self, nums):
        n = len(nums)
        if n == 1: return 1

        dp_length = [1] * n
        dp_count = [1] * n
        max_length = 1
        for i in range(1, n):
            for j in range(i):
                if nums[i] > nums[j]:
                    if dp_length[j] + 1 > dp_length[i]:  # 说明最长递增子序列的长度增加了，dp[i]=dp[j]+1，长度增加，数量不变count[i]=count[j]
                        dp_length[i] = dp_length[j] + 1
                        dp_count[i] = dp_count[j]
                    elif dp_length[j] + 1 == dp_length[i]:  # 说明最长递增子序列的长度并没有增加，但是出现了长度一样的情况，数量增加count[i]+=count[j]
                        dp_count[i] += dp_count[j]
            max_length = max(max_length, dp_length[i])

        res = 0
        for i in range(n):
            if dp_length[i] == max_length:
                res += dp_count[i]
        return res
